source/extensions/access_loggers/wasm/config.cc

```diff
@@ -20,30 +20,29 @@ namespace AccessLoggers {
 namespace Wasm {
 
 AccessLog::InstanceSharedPtr
-WasmAccessLogFactory::createAccessLogInstance(const Protobuf::Message& config,
+WasmAccessLogFactory::createAccessLogInstance(const Protobuf::Message& proto_config,
                                               AccessLog::FilterPtr&& filter,
                                               Server::Configuration::FactoryContext& context) {
-  const auto& wal_config =
-      MessageUtil::downcastAndValidate<const envoy::config::accesslog::v2::WasmAccessLog&>(config);
+  const auto& config =
+      MessageUtil::downcastAndValidate<const envoy::config::accesslog::v2::WasmAccessLog&>(
+          proto_config);
+  auto id = config.id();
+  auto configuration = std::make_shared<std::string>(config.configuration());
   auto tls_slot = context.threadLocal().allocateSlot();
-  if (wal_config.has_vm_config()) {
-    auto base_wasm =
-        Common::Wasm::createWasm(wal_config.id(), wal_config.vm_config(), context.api());
-    tls_slot->set([&wal_config, &base_wasm, &context](Event::Dispatcher& dispatcher) {
-      auto result =
-          Common::Wasm::createThreadLocalWasm(*base_wasm, wal_config.vm_config(), dispatcher,
-                                              wal_config.configuration(), context.api());
-      result->setClusterManager(context.clusterManager());
+  if (config.has_vm_config()) {
+    auto base_wasm = Common::Wasm::createWasm(id, config.vm_config(), context.clusterManager(),
+                                              context.dispatcher(), context.api());
+    // NB: the Slot set() call doesn't complete inline, so all arguments must outlive this call.
+    tls_slot->set([base_wasm, configuration](Event::Dispatcher& dispatcher) {
+      auto result = Common::Wasm::createThreadLocalWasm(*base_wasm, *configuration, dispatcher);
       return std::static_pointer_cast<ThreadLocal::ThreadLocalObject>(result);
     });
   } else {
-    if (wal_config.id().empty()) {
+    if (id.empty()) {
       throw Common::Wasm::WasmVmException("No WASM VM Id or vm_config specified");
     }
-    tls_slot->set([&wal_config, &context](Event::Dispatcher& dispatcher) {
-      auto result = Common::Wasm::getThreadLocalWasm(wal_config.id(), wal_config.configuration());
-      result->setDispatcher(dispatcher);
-      result->setClusterManager(context.clusterManager());
+    tls_slot->set([id, configuration](Event::Dispatcher&) {
+      auto result = Common::Wasm::getThreadLocalWasm(id, *configuration);
       return std::static_pointer_cast<ThreadLocal::ThreadLocalObject>(result);
```

Comment:
`std::shared_ptr<Wasm>` is getting destroyed during the hand-off here, I'm not sure how WAVM survives that.

Instead of using `ThreadLocalObject(std::shared_ptr<Wasm>)`, we could use `ThreadLocalObjectSharedPtr(Wasm)` (i.e. `std::shared_ptr<ThreadLocalObject>(Wasm)`), which is what rest of the codebase seems to be doing.

(same in other extensions)

Reply:
std::static_pointer_cast<> will preserve the std::shared_ptr-ness.  It casts the bit inside the shared_ptr to the new value.  The result of this lambda is ThreadLocalObjectSharedPtr because typedef std::shared_ptr<ThreadLocalObject> ThreadLocalObjectSharedPtr.

Reply:
I do see the destructor being called here (binary built from this branch with no local changes):
```
(lldb) bt
* thread #1, name = 'envoy-static', stop reason = breakpoint 1.2
  * frame #0: 0x0000000000785de4 envoy-static`Envoy::Extensions::Common::Wasm::Wavm::Wavm::~Wavm(this=0x0000000004641200) at wavm.cc:200
    frame #1: 0x000000000077d3cf envoy-static`std::default_delete<Envoy::Extensions::Common::Wasm::WasmVm>::operator(this=0x00000000047c6b10, __ptr=0x0000000004641200)(Envoy::Extensions::Common::Wasm::WasmVm*) const at unique_ptr.h:76
    frame #2: 0x000000000076ef63 envoy-static`std::unique_ptr<Envoy::Extensions::Common::Wasm::WasmVm, std::default_delete<Envoy::Extensions::Common::Wasm::WasmVm> >::~unique_ptr(this=0x00000000047c6b10) at unique_ptr.h:239
    frame #3: 0x0000000000771543 envoy-static`Envoy::Extensions::Common::Wasm::Wasm::~Wasm(this=0x00000000047c6a90) at wasm.h:258
    frame #4: 0x0000000000781cc4 envoy-static`void __gnu_cxx::new_allocator<Envoy::Extensions::Common::Wasm::Wasm>::destroy<Envoy::Extensions::Common::Wasm::Wasm>(this=0x00000000047c6a90, __p=0x00000000047c6a90) at new_allocator.h:124
    frame #5: 0x0000000000781c58 envoy-static`void std::allocator_traits<std::allocator<Envoy::Extensions::Common::Wasm::Wasm> >::destroy<Envoy::Extensions::Common::Wasm::Wasm>(__a=0x00000000047c6a90, __p=0x00000000047c6a90) at alloc_traits.h:487
    frame #6: 0x00000000007819cc envoy-static`std::_Sp_counted_ptr_inplace<Envoy::Extensions::Common::Wasm::Wasm, std::allocator<Envoy::Extensions::Common::Wasm::Wasm>, (__gnu_cxx::_Lock_policy)2>::_M_dispose(this=0x00000000047c6a80) at shared_ptr_base.h:529
    frame #7: 0x000000000047921c envoy-static`std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release(this=0x00000000047c6a80) at shared_ptr_base.h:150
    frame #8: 0x00000000004791ca envoy-static`std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count(this=0x00007fffffffa398) at shared_ptr_base.h:662
    frame #9: 0x00000000004fc739 envoy-static`std::__shared_ptr<Envoy::Extensions::Common::Wasm::Wasm, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr(this=0x00007fffffffa390) at shared_ptr_base.h:928
    frame #10: 0x00000000004fc445 envoy-static`std::shared_ptr<Envoy::Extensions::Common::Wasm::Wasm>::~shared_ptr(this=0x00007fffffffa390) at shared_ptr.h:93
    frame #11: 0x000000000076c623 envoy-static`Envoy::Extensions::Common::Wasm::createWasm(id=(ptr_ = "", length_ = 0), vm_config=0x000000000485ca80, cluster_manager=0x0000000004792f00, dispatcher=0x00000000046343c0, api=0x00000000046ff400) at wasm.cc:1241
    frame #12: 0x000000000076272e envoy-static`Envoy::Extensions::HttpFilters::Wasm::FilterConfig::FilterConfig(this=0x0000000004648800, config=0x000000000485d710, context=0x0000000004793400) at wasm_filter.cc:24
    frame #13: 0x000000000075fbbc envoy-static`Envoy::Extensions::HttpFilters::Wasm::WasmFilterConfig::createFilterFactoryFromProtoTyped(this=0x0000000003c27890, proto_config=0x000000000485d710, (null)="\x04"..., context=0x0000000004793400) at config.cc:20
    frame #14: 0x00000000007604fb envoy-static`Envoy::Extensions::HttpFilters::Common::FactoryBase<envoy::config::filter::http::wasm::v2::Wasm, envoy::config::filter::http::wasm::v2::Wasm>::createFilterFactoryFromProto(this=0x0000000003c27890, proto_config=0x000000000485d710, stats_prefix="\x04"..., context=0x0000000004793400) at factory_base.h:28
    frame #15: 0x0000000001603e06 envoy-static`Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::processFilter(this=0x000000000483b000, proto_config=0x00000000047cbdd0, i=0, prefix=(ptr_ = "http", length_ = 4), filter_factories=size=1)>, std::allocator<std::function<void (Envoy::Http::FilterChainFactoryCallbacks&)> > >&) at config.cc:314
    frame #16: 0x0000000001603037 envoy-static`Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::HttpConnectionManagerConfig(this=0x000000000483b000, config=0x0000000004689320, context=0x0000000004793400, date_provider=0x00000000048666d0, route_config_provider_manager=0x0000000004860830) at config.cc:265
    frame #17: 0x0000000001600b78 envoy-static`Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped(this=0x0000000003c466a0, proto_config=0x0000000004689320, context=0x0000000004793400) at config.cc:91
    frame #18: 0x0000000001608123 envoy-static`Envoy::Extensions::NetworkFilters::Common::FactoryBase<envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager, envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager>::createFilterFactoryFromProto(this=0x0000000003c466a0, proto_config=0x0000000004689320, context=0x0000000004793400) at factory_base.h:28
    frame #19: 0x0000000001a19924 envoy-static`Envoy::Server::ProdListenerComponentFactory::createNetworkFilterFactoryList_(filters=0x00000000046429d0, context=0x0000000004793400) at listener_manager_impl.cc:70
    frame #20: 0x0000000001a2dde8 envoy-static`Envoy::Server::ProdListenerComponentFactory::createNetworkFilterFactoryList(this=0x00000000046b8aa0, filters=0x00000000046429d0, context=0x0000000004793400) at listener_manager_impl.h:57
    frame #21: 0x0000000001a1fbbf envoy-static`Envoy::Server::ListenerImpl::ListenerImpl(this=0x0000000004793400, config=0x00000000046346c0, version_info=", parent=0x00000000046c61c0, name="\b, modifiable=false, workers_started=false, hash=13999327329341325012) at listener_manager_impl.cc:282
    frame #22: 0x0000000001a2411a envoy-static`Envoy::Server::ListenerManagerImpl::addOrUpdateListener(this=0x00000000046c61c0, config=0x00000000046346c0, version_info=", modifiable=false) at listener_manager_impl.cc:746
    frame #23: 0x0000000001a609b3 envoy-static`Envoy::Server::Configuration::MainImpl::initialize(this=0x00000000046b8ae8, bootstrap=0x00000000046b8c38, server=0x00000000046b8a00, cluster_manager_factory=0x0000000004765200) at configuration_impl.cc:68
    frame #24: 0x0000000001a02c97 envoy-static`Envoy::Server::InstanceImpl::initialize(this=0x00000000046b8a00, options=0x00000000046ec338, local_address=const Envoy::Network::Address::Instance @ 0x0000000004626f10, component_factory=0x00000000046ec500) at server.cc:341
    frame #25: 0x00000000019ff9c8 envoy-static`Envoy::Server::InstanceImpl::InstanceImpl(this=0x00000000046b8a00, options=0x00000000046ec338, time_system=0x00000000046ec4e8, local_address=const Envoy::Network::Address::Instance @ 0x0000000004626f10, hooks=0x00000000046ec4f8, restarter=0x00000000046c1200, store=0x000000000463c4e0, access_log_lock=0x00000000046c1240, component_factory=0x00000000046ec500, random_generator=nullptr, tls=0x00000000046feaf0, thread_factory=0x00000000046ec330) at server.cc:85
    frame #26: 0x0000000000495580 envoy-static`std::_MakeUniq<Envoy::Server::InstanceImpl>::__single_object std::make_unique<Envoy::Server::InstanceImpl, Envoy::OptionsImpl&, Envoy::Event::TimeSystem&, std::shared_ptr<Envoy::Network::Address::Instance const>&, Envoy::TestHooks&, Envoy::Server::HotRestart&, Envoy::Stats::ThreadLocalStoreImpl&, Envoy::Thread::BasicLockable&, Envoy::Server::ComponentFactory&, std::unique_ptr<Envoy::Runtime::RandomGenerator, std::default_delete<Envoy::Runtime::RandomGenerator> >, Envoy::ThreadLocal::InstanceImpl&, Envoy::Thread::ThreadFactory&>(__args=0x00000000046ec338, __args=0x00000000046ec4e8, __args=const Envoy::Network::Address::Instance @ 0x0000000004626f10, __args=0x00000000046ec4f8, __args=0x00000000046c1200, __args=0x000000000463c4e0, __args=0x00000000046c1240, __args=0x00000000046ec500, __args=nullptr, __args=0x00000000046feaf0, __args=0x00000000046ec330) at unique_ptr.h:791
    frame #27: 0x0000000000493446 envoy-static`Envoy::MainCommonBase::MainCommonBase(this=0x00000000046ec508, options=0x00000000046ec338, time_system=0x00000000046ec4e8, test_hooks=0x00000000046ec4f8, component_factory=0x00000000046ec500, random_generator=nullptr, thread_factory=0x00000000046ec330) at main_common.cc:82
    frame #28: 0x0000000000494028 envoy-static`Envoy::MainCommon::MainCommon(this=0x00000000046ec000, argc=7, argv=0x00007fffffffe408) at main_common.cc:139
    frame #29: 0x0000000000476f50 envoy-static`std::_MakeUniq<Envoy::MainCommon>::__single_object std::make_unique<Envoy::MainCommon, int&, char**&>(__args=0x00007fffffffe318, __args=0x00007fffffffe310) at unique_ptr.h:791
    frame #30: 0x00000000004769bf envoy-static`main(argc=7, argv=0x00007fffffffe408) at main.cc:26
    frame #31: 0x00007ffff712d2e1 libc.so.6`__libc_start_main(main=(envoy-static`main at main.cc:14), argc=7, argv=0x00007fffffffe408, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffe3f8) at libc-start.c:291
    frame #32: 0x00000000004768aa envoy-static`_start + 42
```

Reply:
That is bizarre.  The docs say otherwise: https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast

AFAICT this is coming out of  wasm.cc:1241 which is the return statement in createWasm which is just returning a std::shared_ptr<Wasm> from the stack.

I am not sure what to make of the stack trace.  It doesn't seem to be coming from the std::static_pointer_cast location, but conversely it shouldn't be calling the destructor because it is just returning a shared_ptr.

Reply:
The std::static_pointer_cast isn't going to appear in the stack trace because those calls are occurring on the silo threads.  It is true that if the thread callbacks occur before the end of the scope because of scheduling issues then when base_wasm goes out of scope it will be the last shared_ptr and will result in the Wasm being deleted.... could that be what we are seeing because of some optimizations in envoy-static?

Reply:
wasm.cc:1241 is only called when createWasm is used to build the base_wasm.  This is then passed by value into the lamba,  The delete could happen if (for example) there where no threads such that tls_slot_->set() return immediately.

Reply:
OK I tested this and it is as I suspected.   If you set a breakpoints at 

1) ~Wasm
2) tls_slot_->set([base_wasm, configuration](Event::Dispatcher& dispatcher) { 
3) inside the lambda at createThreadLocalWasm

you will see

2, 3, 3, 3, 3, 3, ... 3, 1

This is expected.  The base_wasm goes out of scope after the lambdas have already destroyed their local copies of the shard_ptr after making clones (for WAVM)

Reply:
Ah, this is the `base_wasm` indeed. My fault for testing with single worker, which didn't make this more obvious.

I got confused by the fact that the destructor is called on the worker thread, and not on the main thread, which created `base_wasm`.

Thanks for digging into this!